### PR TITLE
Automatic jbuild --> dune upgrader

### PR DESCRIPTION
**Pull request by ghost**
This PR adds a `dune upgrade` command that automatically upgrade all the jbuilder projects in the workspace to dune.

Limitations:
- it cannot upgrade `jbuild` files written in OCaml syntax

### CODE REVIEW

**Comment by diml** [comment]
This is ready for review. I tested it and it seems to do the job. `dune upgrade` takes care of calling the right `git` commands so that the user can do `git diff` just after `dune upgrade` and see what changed.

To be properly usable, we will need to finish `dune unstable-fmt` so that it preserves comments and string layouts.

**Comment by diml** [comment]
As a side note, I also modified the behaviour so that when we automatically create the `dune-project` file we also insert the detected project name.

**Comment by rgrinberg** [comment]
> As a side note, I also modified the behaviour so that when we automatically create the dune-project file we also insert the detected project name.

I think this deserves a CHANGES entry as well.

**Comment by diml** [comment]
Indeed, I added one.

File: src/upgrader.ml [hunk 1]
```diff
@@ -0,0 +1,243 @@
...
+    } :: todo.to_rename_and_edit
+
+let upgrade_dir todo dir =
+  let project = File_tree.Dir.project dir in
+  (match Dune_project.ensure_project_file_exists project with
+   | Already_exist -> ()
+   | Created ->
+     let project_file =
+       Path.relative (Path.of_local (Dune_project.root project))
+         Dune_project.filename
```
Comment (rgrinberg) [inline src/upgrader.ml]: It might simpler to have a `Dune_project.path : t -> Path.t` function that returns the path of the dune project.

**Review by rgrinberg — APPROVED** [review]
Looks good. A test wouldn't hurt either.

File: src/upgrader.ml (hunk 1 above)
Comment (diml) [reply src/upgrader.ml]: Sounds good, I name it just `Dune_project.file`.

**Comment by diml** [comment]
> Looks good. A test wouldn't hurt either.

True, but at the same time this feature will only exist for the next 6 months, so I'll be lazy here :)

**Comment by emillon** [comment]
Thanks! It will exist for next 6 months, but at the same time the generated code will be around for longer :)

**Comment by rgrinberg** [comment]
Btw, one good test case would be to upgrade all of our blackbox tests that use jbuild files

**Pull request merged**